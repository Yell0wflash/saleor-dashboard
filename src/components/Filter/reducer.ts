import { update } from "@saleor/utils/lists";
import { IFilter, IFilterElementMutableData } from "./types";

export type FilterReducerActionType = "clear" | "reset" | "set-property";
export interface FilterReducerAction<T extends string> {
  type: FilterReducerActionType;
  payload: Partial<{
    name: T;
    update: Partial<IFilterElementMutableData>;
    reset: IFilter<T>;
  }>;
}

function setProperty<T extends string>(
  prevState: IFilter<T>,
  filter: T,
  updateData: Partial<IFilterElementMutableData>
): IFilter<T> {
  const field = prevState.find(f => f.name === filter);
  const updatedField = {
    ...field,
    ...updateData
  };

  return update(updatedField, prevState, (a, b) => a.name === b.name);
}

function reduceFilter<T extends string>(
  prevState: IFilter<T>,
  action: FilterReducerAction<T>
): IFilter<T> {
  switch (action.type) {
    case "clear":
      return prevState;
    case "set-property":
      return setProperty(prevState, action.payload.name, action.payload.update);
    case "reset":
      return action.payload.reset;
  }
  return prevState;
}

export default reduceFilter;
