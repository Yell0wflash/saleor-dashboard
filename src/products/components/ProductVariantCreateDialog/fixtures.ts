import { initialForm, ProductVariantCreateFormData } from "./form";

export const attributes = [
  {
    id: "attr-1",
    values: Array(9)
      .fill(0)
      .map((_, index) => `val-1-${index + 1}`)
  },
  {
    id: "attr-2",
    values: Array(6)
      .fill(0)
      .map((_, index) => `val-2-${index + 1}`)
  },
  {
    id: "attr-3",
    values: Array(4)
      .fill(0)
      .map((_, index) => `val-3-${index + 1}`)
  },
  {
    id: "attr-4",
    values: Array(11)
      .fill(0)
      .map((_, index) => `val-4-${index + 1}`)
  }
];

export const secondStep: ProductVariantCreateFormData = {
  ...initialForm,
  attributes: [
    {
      id: attributes[0].id,
      values: []
    },
    {
      id: attributes[1].id,
      values: []
    },
    {
      id: attributes[3].id,
      values: []
    }
  ]
};
export const thirdStep: ProductVariantCreateFormData = {
  ...secondStep,
  attributes: [
    {
      id: attributes[0].id,
      values: [0, 6].map(index => attributes[0].values[index])
    },
    {
      id: attributes[1].id,
      values: [1, 3].map(index => attributes[1].values[index])
    },
    {
      id: attributes[3].id,
      values: [0, 4].map(index => attributes[3].values[index])
    }
  ]
};
