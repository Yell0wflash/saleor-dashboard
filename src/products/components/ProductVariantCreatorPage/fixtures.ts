import { createVariants } from "./createVariants";
import {
  AllOrAttribute,
  createInitialForm,
  ProductVariantCreateFormData
} from "./form";

export const attributes = [
  {
    id: "attr-1",
    values: Array(9)
      .fill(0)
      .map((_, index) => `val-1-${index + 1}`)
  },
  {
    id: "attr-2",
    values: Array(6)
      .fill(0)
      .map((_, index) => `val-2-${index + 1}`)
  },
  {
    id: "attr-3",
    values: Array(4)
      .fill(0)
      .map((_, index) => `val-3-${index + 1}`)
  },
  {
    id: "attr-4",
    values: Array(11)
      .fill(0)
      .map((_, index) => `val-4-${index + 1}`)
  }
];

export const secondStep: ProductVariantCreateFormData = {
  ...createInitialForm([], "10.99"),
  attributes: [
    {
      id: attributes[0].id,
      values: []
    },
    {
      id: attributes[1].id,
      values: []
    },
    {
      id: attributes[3].id,
      values: []
    }
  ]
};

export const thirdStep: ProductVariantCreateFormData = {
  ...secondStep,
  attributes: [
    {
      id: attributes[0].id,
      values: [0, 6].map(index => attributes[0].values[index])
    },
    {
      id: attributes[1].id,
      values: [1, 3].map(index => attributes[1].values[index])
    },
    {
      id: attributes[3].id,
      values: [0, 4].map(index => attributes[3].values[index])
    }
  ]
};

const price: AllOrAttribute = {
  all: false,
  attribute: thirdStep.attributes[1].id,
  value: "",
  values: [
    {
      slug: thirdStep.attributes[1].values[0],
      value: "24.99"
    },
    {
      slug: thirdStep.attributes[1].values[1],
      value: "26.99"
    }
  ]
};
const stock: AllOrAttribute = {
  all: false,
  attribute: thirdStep.attributes[2].id,
  value: "",
  values: [
    {
      slug: thirdStep.attributes[2].values[0],
      value: "50"
    },
    {
      slug: thirdStep.attributes[2].values[1],
      value: "35"
    }
  ]
};
export const fourthStep: ProductVariantCreateFormData = {
  ...thirdStep,
  price,
  stock,
  variants: createVariants({
    ...thirdStep,
    price,
    stock
  })
};
