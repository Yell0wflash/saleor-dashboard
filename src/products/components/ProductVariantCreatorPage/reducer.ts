import {
  add,
  remove,
  removeAtIndex,
  toggle,
  updateAtIndex
} from "@saleor/utils/lists";
import { createVariants } from "./createVariants";
import { ProductVariantCreateFormData } from "./form";

export type ProductVariantCreateReducerActionType =
  | "applyPriceToAll"
  | "applyPriceToAttribute"
  | "applyStockToAll"
  | "applyStockToAttribute"
  | "changeApplyPriceToAllValue"
  | "changeApplyPriceToAttributeId"
  | "changeApplyStockToAllValue"
  | "changeApplyStockToAttributeId"
  | "changeAttributeValuePrice"
  | "changeAttributeValueStock"
  | "changeVariantData"
  | "deleteVariant"
  | "reload"
  | "selectValue";

export type VariantField = "stock" | "price" | "sku";
export interface ProductVariantCreateReducerAction {
  all?: boolean;
  attributeId?: string;
  data?: ProductVariantCreateFormData;
  field?: VariantField;
  type: ProductVariantCreateReducerActionType;
  value?: string;
  valueId?: string;
  variantIndex?: number;
}

function selectValue(
  prevState: ProductVariantCreateFormData,
  attributeId: string,
  valueSlug: string
): ProductVariantCreateFormData {
  const attribute = prevState.attributes.find(
    attribute => attribute.id === attributeId
  );
  const values = toggle(valueSlug, attribute.values, (a, b) => a === b);
  const updatedAttributes = add(
    {
      id: attributeId,
      values
    },
    remove(attribute, prevState.attributes, (a, b) => a.id === b.id)
  );

  const priceValues =
    prevState.price.attribute === attributeId
      ? toggle(
          {
            slug: valueSlug,
            value: ""
          },
          prevState.price.values,
          (a, b) => a.slug === b.slug
        )
      : prevState.price.values;

  const stockValues =
    prevState.stock.attribute === attributeId
      ? toggle(
          {
            slug: valueSlug,
            value: ""
          },
          prevState.stock.values,
          (a, b) => a.slug === b.slug
        )
      : prevState.stock.values;

  return {
    ...prevState,
    attributes: updatedAttributes,
    price: {
      ...prevState.price,
      values: priceValues
    },
    stock: {
      ...prevState.stock,
      values: stockValues
    }
  };
}

function applyPriceToAll(
  state: ProductVariantCreateFormData,
  value: boolean
): ProductVariantCreateFormData {
  const data = {
    ...state,
    price: {
      ...state.price,
      all: value
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function applyStockToAll(
  state: ProductVariantCreateFormData,
  value: boolean
): ProductVariantCreateFormData {
  const data = {
    ...state,
    stock: {
      ...state.stock,
      all: value
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeAttributeValuePrice(
  state: ProductVariantCreateFormData,
  attributeValueSlug: string,
  price: string
): ProductVariantCreateFormData {
  const index = state.price.values.findIndex(
    value => value.slug === attributeValueSlug
  );

  if (index === -1) {
    throw new Error(`Value with id ${attributeValueSlug} not found`);
  }

  const values = updateAtIndex(
    {
      slug: attributeValueSlug,
      value: price
    },
    state.price.values,
    index
  );

  const data = {
    ...state,
    price: {
      ...state.price,
      values
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeAttributeValueStock(
  state: ProductVariantCreateFormData,
  attributeValueSlug: string,
  stock: string
): ProductVariantCreateFormData {
  const index = state.stock.values.findIndex(
    value => value.slug === attributeValueSlug
  );

  if (index === -1) {
    throw new Error(`Value with id ${attributeValueSlug} not found`);
  }

  const values = updateAtIndex(
    {
      slug: attributeValueSlug,
      value: stock
    },
    state.stock.values,
    index
  );

  const data = {
    ...state,
    stock: {
      ...state.stock,
      values
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeApplyPriceToAttributeId(
  state: ProductVariantCreateFormData,
  attributeId: string
): ProductVariantCreateFormData {
  const attribute = state.attributes.find(
    attribute => attribute.id === attributeId
  );
  const values = attribute.values.map(slug => ({
    slug,
    value: ""
  }));
  const data = {
    ...state,
    price: {
      ...state.price,
      attribute: attributeId,
      values
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeApplyStockToAttributeId(
  state: ProductVariantCreateFormData,
  attributeId: string
): ProductVariantCreateFormData {
  const attribute = state.attributes.find(
    attribute => attribute.id === attributeId
  );
  const values = attribute.values.map(slug => ({
    slug,
    value: ""
  }));

  const data = {
    ...state,
    stock: {
      ...state.stock,
      attribute: attributeId,
      values
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeApplyPriceToAllValue(
  state: ProductVariantCreateFormData,
  value: string
): ProductVariantCreateFormData {
  const data = {
    ...state,
    price: {
      ...state.price,
      value
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeApplyStockToAllValue(
  state: ProductVariantCreateFormData,
  value: string
): ProductVariantCreateFormData {
  const data = {
    ...state,
    stock: {
      ...state.stock,
      value
    }
  };

  return {
    ...data,
    variants: createVariants(data)
  };
}

function changeVariantData(
  state: ProductVariantCreateFormData,
  field: VariantField,
  value: string,
  variantIndex: number
): ProductVariantCreateFormData {
  const variant = state.variants[variantIndex];
  if (field === "price") {
    variant.priceOverride = value;
  } else if (field === "sku") {
    variant.sku = value;
  } else {
    variant.quantity = parseInt(value, 10);
  }

  return {
    ...state,
    variants: updateAtIndex(variant, state.variants, variantIndex)
  };
}

function deleteVariant(
  state: ProductVariantCreateFormData,
  variantIndex: number
): ProductVariantCreateFormData {
  return {
    ...state,
    variants: removeAtIndex(state.variants, variantIndex)
  };
}

function reduceProductVariantCreateFormData(
  prevState: ProductVariantCreateFormData,
  action: ProductVariantCreateReducerAction
) {
  switch (action.type) {
    case "selectValue":
      return selectValue(prevState, action.attributeId, action.valueId);

    case "applyPriceToAll":
      return applyPriceToAll(prevState, action.all);
    case "applyStockToAll":
      return applyStockToAll(prevState, action.all);
    case "changeAttributeValuePrice":
      return changeAttributeValuePrice(prevState, action.valueId, action.value);
    case "changeAttributeValueStock":
      return changeAttributeValueStock(prevState, action.valueId, action.value);
    case "changeApplyPriceToAttributeId":
      return changeApplyPriceToAttributeId(prevState, action.attributeId);
    case "changeApplyStockToAttributeId":
      return changeApplyStockToAttributeId(prevState, action.attributeId);
    case "changeApplyPriceToAllValue":
      return changeApplyPriceToAllValue(prevState, action.value);
    case "changeApplyStockToAllValue":
      return changeApplyStockToAllValue(prevState, action.value);
    case "changeVariantData":
      return changeVariantData(
        prevState,
        action.field,
        action.value,
        action.variantIndex
      );
    case "deleteVariant":
      return deleteVariant(prevState, action.variantIndex);
    case "reload":
      return action.data;
    default:
      return prevState;
  }
}

export default reduceProductVariantCreateFormData;
