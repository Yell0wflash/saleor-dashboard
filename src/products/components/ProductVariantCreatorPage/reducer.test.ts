import { attributes, fourthStep, secondStep, thirdStep } from "./fixtures";
import reducer, { VariantField } from "./reducer";

function execActions<TState, TAction>(
  initialState: TState,
  reducer: (state: TState, action: TAction) => TState,
  actions: TAction[]
): TState {
  return actions.reduce((acc, action) => reducer(acc, action), initialState);
}

describe("Reducer is able to", () => {
  it("select attribute values", () => {
    const state = execActions(secondStep, reducer, [
      {
        attributeId: attributes[0].id,
        type: "selectValue",
        valueId: attributes[0].values[0]
      },
      {
        attributeId: attributes[0].id,
        type: "selectValue",
        valueId: attributes[0].values[6]
      },
      {
        attributeId: attributes[1].id,
        type: "selectValue",
        valueId: attributes[1].values[1]
      },
      {
        attributeId: attributes[1].id,
        type: "selectValue",
        valueId: attributes[1].values[3]
      },
      {
        attributeId: attributes[3].id,
        type: "selectValue",
        valueId: attributes[3].values[0]
      },
      {
        attributeId: attributes[3].id,
        type: "selectValue",
        valueId: attributes[3].values[4]
      }
    ]);

    expect(state.attributes[0].values).toHaveLength(2);
    expect(state.attributes[1].values).toHaveLength(2);
    expect(state.attributes[2].values).toHaveLength(2);
    expect(state).toMatchSnapshot();
  });

  it("select price for all variants", () => {
    const value = "45.99";
    const state = execActions(thirdStep, reducer, [
      {
        all: true,
        type: "applyPriceToAll"
      },
      {
        type: "changeApplyPriceToAllValue",
        value
      }
    ]);

    expect(state.price.all).toBeTruthy();
    expect(state.price.value).toBe(value);
    expect(state).toMatchSnapshot();
  });

  it("select stock for all variants", () => {
    const value = 45.99;
    const state = execActions(thirdStep, reducer, [
      {
        all: true,
        type: "applyStockToAll"
      },
      {
        type: "changeApplyStockToAllValue",
        value: value.toString()
      }
    ]);

    expect(state.stock.all).toBeTruthy();
    expect(state.stock.value).toBe(value.toString());
    expect(state).toMatchSnapshot();
  });

  it("select price to each attribute value", () => {
    const attribute = thirdStep.attributes[0];
    const value = 45.99;
    const state = execActions(thirdStep, reducer, [
      {
        all: false,
        type: "applyPriceToAll"
      },
      {
        attributeId: attribute.id,
        type: "changeApplyPriceToAttributeId"
      },
      {
        type: "changeAttributeValuePrice",
        value: value.toString(),
        valueId: attribute.values[0]
      },
      {
        type: "changeAttributeValuePrice",
        value: (value + 6).toString(),
        valueId: attribute.values[1]
      }
    ]);

    expect(state.price.all).toBeFalsy();
    expect(state.price.values).toHaveLength(
      state.attributes.find(attribute => state.price.attribute === attribute.id)
        .values.length
    );
    expect(state).toMatchSnapshot();
  });

  it("select stock to each attribute value", () => {
    const attribute = thirdStep.attributes[0];
    const value = 13;
    const state = execActions(thirdStep, reducer, [
      {
        all: false,
        type: "applyStockToAll"
      },
      {
        attributeId: attribute.id,
        type: "changeApplyStockToAttributeId"
      },
      {
        type: "changeAttributeValueStock",
        value: value.toString(),
        valueId: attribute.values[0]
      },
      {
        type: "changeAttributeValueStock",
        value: (value + 6).toString(),
        valueId: attribute.values[1]
      }
    ]);

    expect(state.stock.all).toBeFalsy();
    expect(state.stock.values).toHaveLength(
      state.attributes.find(attribute => state.stock.attribute === attribute.id)
        .values.length
    );
    expect(state).toMatchSnapshot();
  });

  it("modify individual variant price", () => {
    const field: VariantField = "price";
    const value = "49.99";
    const variantIndex = 3;

    const state = execActions(fourthStep, reducer, [
      {
        field,
        type: "changeVariantData",
        value,
        variantIndex
      }
    ]);

    expect(state.variants[variantIndex].priceOverride).toBe(value);
    expect(state.variants[variantIndex - 1].priceOverride).toBe(
      fourthStep.variants[variantIndex - 1].priceOverride
    );
  });

  it("delete variant", () => {
    const variantIndex = 3;

    const state = execActions(fourthStep, reducer, [
      {
        type: "deleteVariant",
        variantIndex
      }
    ]);

    expect(state.variants.length).toBe(fourthStep.variants.length - 1);
  });
});
