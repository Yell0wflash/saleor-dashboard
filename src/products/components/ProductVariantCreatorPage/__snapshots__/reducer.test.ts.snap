// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer is able to select attribute values 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "price": Object {
    "all": true,
    "attribute": undefined,
    "value": "10.99",
    "values": Array [],
  },
  "stock": Object {
    "all": true,
    "attribute": undefined,
    "value": "",
    "values": Array [],
  },
  "variants": Array [],
}
`;

exports[`Reducer is able to select price for all variants 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "price": Object {
    "all": true,
    "attribute": undefined,
    "value": "45.99",
    "values": Array [],
  },
  "stock": Object {
    "all": true,
    "attribute": undefined,
    "value": "",
    "values": Array [],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
  ],
}
`;

exports[`Reducer is able to select price to each attribute value 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "price": Object {
    "all": false,
    "attribute": "attr-1",
    "value": "10.99",
    "values": Array [
      Object {
        "slug": "val-1-1",
        "value": "45.99",
      },
      Object {
        "slug": "val-1-7",
        "value": "51.99",
      },
    ],
  },
  "stock": Object {
    "all": true,
    "attribute": undefined,
    "value": "",
    "values": Array [],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "45.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "51.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "51.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "51.99",
      "quantity": NaN,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "51.99",
      "quantity": NaN,
      "sku": "",
    },
  ],
}
`;

exports[`Reducer is able to select stock for all variants 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "price": Object {
    "all": true,
    "attribute": undefined,
    "value": "10.99",
    "values": Array [],
  },
  "stock": Object {
    "all": true,
    "attribute": undefined,
    "value": "45.99",
    "values": Array [],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 45,
      "sku": "",
    },
  ],
}
`;

exports[`Reducer is able to select stock to each attribute value 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "price": Object {
    "all": true,
    "attribute": undefined,
    "value": "10.99",
    "values": Array [],
  },
  "stock": Object {
    "all": false,
    "attribute": "attr-1",
    "value": "",
    "values": Array [
      Object {
        "slug": "val-1-1",
        "value": "13",
      },
      Object {
        "slug": "val-1-7",
        "value": "19",
      },
    ],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 13,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 13,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 13,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 13,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 19,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 19,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 19,
      "sku": "",
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "priceOverride": "10.99",
      "quantity": 19,
      "sku": "",
    },
  ],
}
`;
